lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+
+  auto intfData = m_pipelineState->getShaderInterfaceData(m_shaderStage);
+  auto resUsage = m_pipelineState->getShaderResourceUsage(m_shaderStage);
   auto &entryArgIdxs = intfData->entryArgIdxs;
-  entryArgIdxs.initialized = true;
+  auto &builtInUsage = resUsage->builtInUsage;
+  bool enableMultiView = m_pipelineState->getInputAssemblyState().enableMultiView;
+  bool enableNgg = m_pipelineState->isGraphics() ? m_pipelineState->getNggControl()->enableNgg : false;
 
-  // Estimated available user data count
-  unsigned maxUserDataCount = m_pipelineState->getTargetInfo().getGpuProperty().maxUserDataCount;
-  unsigned availUserDataCount = maxUserDataCount - userDataIdx;
-  unsigned requiredRemappedUserDataCount = 0; // Maximum required user data
-  unsigned requiredUserDataCount = 0;         // Maximum required user data without remapping
-  bool useFixedLayout = m_shaderStage == ShaderStageCompute && !m_pipelineState->isUnlinked();
-  bool reserveVbTable = false;
+  // See if there is a vertex buffer and/or streamout in the root user data layout.
+  bool reserveVertexBufferTable = false;
   bool reserveStreamOutTable = false;
-  bool reserveEsGsLdsSize = false;
-  bool needSpill = false;
-
-  if (userDataNodes.size() > 0) {
-    for (unsigned i = 0; i < userDataNodes.size(); ++i) {
-      auto node = &userDataNodes[i];
-      // NOTE: Per PAL request, the value of IndirectTableEntry is the node offset + 1.
-      // and indirect user data should not be counted in possible spilled user data.
-      if (node->type == ResourceNodeType::IndirectUserDataVaPtr) {
-        // Only the vertex shader needs a vertex buffer table.
-        if (m_shaderStage == ShaderStageVertex)
-          reserveVbTable = true;
-        else if (m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9) {
-          // On GFX9+, the shader stage that the vertex shader is merged in to needs a vertex buffer
-          // table, to ensure that the merged shader gets one.
-          if (m_shaderStage == ShaderStageTessControl || (m_shaderStage == ShaderStageGeometry && !m_hasTs))
-            reserveVbTable = true;
-        }
-        continue;
-      }
-
-      if (node->type == ResourceNodeType::StreamOutTableVaPtr) {
-        // Only the last shader stage before fragment (ignoring copy shader) needs a stream out table.
-        if ((m_pipelineState->getShaderStageMask() &
-             (shaderStageToMask(ShaderStageFragment) - shaderStageToMask(m_shaderStage))) ==
-            shaderStageToMask(m_shaderStage))
+  for (auto &node : m_pipelineState->getUserDataNodes()) {
+    if (node.type == ResourceNodeType::IndirectUserDataVaPtr) {
+      reserveVertexBufferTable = true;
+    } else if (node.type == ResourceNodeType::StreamOutTableVaPtr) {
+      // Only the last shader stage before fragment (ignoring copy shader) needs a stream out table.
+      if ((m_pipelineState->getShaderStageMask() &
```

Comment:
Could you help to add more comment for the stream out stage check? this code is too tricky.

Reply:
I changed it to use PipelineState::getLastVertexProcessingState, a method I had already created in an earlier change.